File: contests/lc/3531.CountCoveredBuildings/solution.py
from typing import List


class Solution:
    def countCoveredBuildings(self, n: int, buildings: List[List[int]]) -> int:
        # Track the smallest and largest column per row, and row per column.
        row_limits: dict[int, tuple[int, int]] = {}
        col_limits: dict[int, tuple[int, int]] = {}

        for r, c in buildings:
            if r in row_limits:
                row_min, row_max = row_limits[r]
                row_limits[r] = (min(row_min, c), max(row_max, c))
            else:
                row_limits[r] = (c, c)

            if c in col_limits:
                col_min, col_max = col_limits[c]
                col_limits[c] = (min(col_min, r), max(col_max, r))
            else:
                col_limits[c] = (r, r)

        covered = 0
        for r, c in buildings:
            row_min, row_max = row_limits[r]
            col_min, col_max = col_limits[c]

            if c != row_min and c != row_max and r != col_min and r != col_max:
                covered += 1

        return covered

File: contests/lc/3531.CountCoveredBuildings/solution.go
package countcoveredbuildings

// countCoveredBuildings returns the number of buildings that have at least
// one building in each cardinal direction (left, right, above, below).
func countCoveredBuildings(n int, buildings [][]int) int {
	// Track smallest and largest column per row, and row per column.
	rowLimits := make(map[int][2]int)
	colLimits := make(map[int][2]int)

	for _, b := range buildings {
		r, c := b[0], b[1]

		if lim, ok := rowLimits[r]; ok {
			rowLimits[r] = [2]int{min(lim[0], c), max(lim[1], c)}
		} else {
			rowLimits[r] = [2]int{c, c}
		}

		if lim, ok := colLimits[c]; ok {
			colLimits[c] = [2]int{min(lim[0], r), max(lim[1], r)}
		} else {
			colLimits[c] = [2]int{r, r}
		}
	}

	covered := 0
	for _, b := range buildings {
		r, c := b[0], b[1]
		rowMin, rowMax := rowLimits[r][0], rowLimits[r][1]
		colMin, colMax := colLimits[c][0], colLimits[c][1]

		if c != rowMin && c != rowMax && r != colMin && r != colMax {
			covered++
		}
	}

	return covered
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
